src/java/org/apache/cassandra/index/sai/plan/Plan.java

```diff
@@ -296,13 +295,34 @@ public final String toStringRecursive()
         return formatter.format(this);
     }
 
+    /**
+     * Formats the whole plan as a pretty tree with redacted information for slow query logging
+     */
+    public final String toStringRecursiveRedacted()
+    {
+        TreeFormatter<Plan> formatter = new TreeFormatter<>(Plan::toStringRedacted, Plan::subplans);
+        return formatter.format(this);
+    }
+
     /**
      * Returns the string representation of this node only
      */
     public final String toString()
     {
-        String title = title();
-        String description = description();
+        String title = title(false);
+        String description = description(false);
+        return (title.isEmpty())
+               ? String.format("%s (%s)\n%s", getClass().getSimpleName(), cost(), description).stripTrailing()
+               : String.format("%s %s (%s)\n%s", getClass().getSimpleName(), title, cost(), description).stripTrailing();
+    }
+
+    /**
+     * Returns the redacted string representation of this node only (used in slow query logging)
+     */
+    public final String toStringRedacted()
```

Comment:
I should add here `toStringInternal` again, in order to minimize the repetition of code

Reply:
I don't understand what you are saying here.
Do you say even the current `toStringRecursive `should be redacted? It seems we print the trees only in trace logs and we never enable trace in prod in CNDB. Though, I am not sure it is not a problem for HCD. I think it is a good idea to ask Pushkala. Here, we log values even from prepared statements.

`toStringPretty(FormattingOptions options)`
What would be `FormattingOptions`? I think I am not sure what you have in mind.